handle empty release in rpmquery.py

lethliel opened the pull request:
This is very unlikely but in very rare cases this
can happen.

Builing kiwi images containing debian is one case.
In this case we do not know what is inside the kiwi
file and osc build assumes buildtype 'rpm' to generate
a package list which get's thrown away anyway.

Now we just check for release

lethliel commented:
@marcus-h: Please review and merge this also

marcus-h commented:
On 2019-10-21 02:11:28 -0700, Marco Strigl wrote:
> This is very unlikely but in very rare cases this
> can happen.
>
> Builing kiwi images containing debian is one case.
> In this case we do not know what is inside the kiwi
> file and osc build assumes buildtype &#39;rpm&#39; to generate
> a package list which get&#39;s thrown away anyway.
>
Hmm do you have a concrete example where this happens? Would it
help if we "fix" the caller to provide a dummy value for the release?

(I would like to avoid such a change in the rpmquery because for a
plain rpm, a release always exists.)

lethliel commented:
On 10/21/19 4:10 PM, Marcus Hüwe wrote:
> On 2019-10-21 02:11:28 -0700, Marco Strigl wrote:
>  > This is very unlikely but in very rare cases this
>  > can happen.
>  >
>  > Builing kiwi images containing debian is one case.
>  > In this case we do not know what is inside the kiwi
>  > file and osc build assumes buildtype &#39;rpm&#39; to generate
>  > a package list which get&#39;s thrown away anyway.
>  >
> Hmm do you have a concrete example where this happens? Would it
> help if we "fix" the caller to provide a dummy value for the release?
> 
> (I would like to avoid such a change in the rpmquery because for a
> plain rpm, a release always exists.)
> 

This happens for devel:kanku:images/ubuntu-JeOS:

`osc build images_ubuntu`

creates:

```

Building ubuntu-JeOS.kiwi for images_ubuntu/x86_64
Getting buildinfo from server and store to 
/home/frank/obs/devel:kanku:images/
ubuntu-JeOS/.osc/_buildinfo-images_ubuntu-x86_64.xml
Traceback (most recent call last):
   File "/usr/bin/osc", line 41, in <module>
     r = babysitter.run(osccli)
   File "/usr/lib/python3.7/site-packages/osc/babysitter.py", line 62, 
in run
     return prg.main(argv)
   File "/usr/lib/python3.7/site-packages/osc/cmdln.py", line 344, in main
     return self.cmd(args)
   File "/usr/lib/python3.7/site-packages/osc/cmdln.py", line 367, in cmd
     retval = self.onecmd(argv)
   File "/usr/lib/python3.7/site-packages/osc/cmdln.py", line 501, in onecmd
     return self._dispatch_cmd(handler, argv)
   File "/usr/lib/python3.7/site-packages/osc/cmdln.py", line 1232, in
_dispatch_cmd
     return handler(argv[0], opts, *args)
   File "/usr/lib/python3.7/site-packages/osc/commandline.py", line 
6399, in
do_build
     return osc.build.main(self.get_api_url(), opts, args)
   File "/usr/lib/python3.7/site-packages/osc/build.py", line 819, in main
     kiwipath = get_kiwipath_from_buildinfo(apiurl, bi_filename, prj, repo)
   File "/usr/lib/python3.7/site-packages/osc/build.py", line 518, in
get_kiwipath_from_buildinfo
     bi = Buildinfo(bi_filename, apiurl, 'kiwi')
   File "/usr/lib/python3.7/site-packages/osc/build.py", line 159, in 
__init__
     apiurl, localpkgs)
   File "/usr/lib/python3.7/site-packages/osc/build.py", line 249, in 
__init__
     canonname = rpmquery.RpmQuery.filename(self.mp['name'].encode(), 
epoch,
self.mp['version'].encode(), release, self.mp['arch'].encode())
   File "/usr/lib/python3.7/site-packages/osc/util/rpmquery.py", line 
377, in
filename
     return b'%s-%s-%s.%s.rpm' % (name, version, release, arch)
TypeError: %b requires a bytes-like object, or an object that implements
__bytes__, not 'NoneType'
```

lethliel commented:
for devel:kanku:images/ubuntu-JeOS

using
`osc build images_ubuntu`

this is thrown: 
```
Building ubuntu-JeOS.kiwi for images_ubuntu/x86_64
Getting buildinfo from server and store to /home/frank/obs/devel:kanku:images/
ubuntu-JeOS/.osc/_buildinfo-images_ubuntu-x86_64.xml
Traceback (most recent call last):
  File "/usr/bin/osc", line 41, in <module>
    r = babysitter.run(osccli)
  File "/usr/lib/python3.7/site-packages/osc/babysitter.py", line 62, in run
    return prg.main(argv)
  File "/usr/lib/python3.7/site-packages/osc/cmdln.py", line 344, in main
    return self.cmd(args)
  File "/usr/lib/python3.7/site-packages/osc/cmdln.py", line 367, in cmd
    retval = self.onecmd(argv)
  File "/usr/lib/python3.7/site-packages/osc/cmdln.py", line 501, in onecmd
    return self._dispatch_cmd(handler, argv)
  File "/usr/lib/python3.7/site-packages/osc/cmdln.py", line 1232, in 
_dispatch_cmd
    return handler(argv[0], opts, *args)
  File "/usr/lib/python3.7/site-packages/osc/commandline.py", line 6399, in 
do_build
    return osc.build.main(self.get_api_url(), opts, args)
  File "/usr/lib/python3.7/site-packages/osc/build.py", line 819, in main
    kiwipath = get_kiwipath_from_buildinfo(apiurl, bi_filename, prj, repo)
  File "/usr/lib/python3.7/site-packages/osc/build.py", line 518, in 
get_kiwipath_from_buildinfo
    bi = Buildinfo(bi_filename, apiurl, 'kiwi')
  File "/usr/lib/python3.7/site-packages/osc/build.py", line 159, in __init__
    apiurl, localpkgs)
  File "/usr/lib/python3.7/site-packages/osc/build.py", line 249, in __init__
    canonname = rpmquery.RpmQuery.filename(self.mp['name'].encode(), epoch, 
self.mp['version'].encode(), release, self.mp['arch'].encode())
  File "/usr/lib/python3.7/site-packages/osc/util/rpmquery.py", line 377, in 
filename
    return b'%s-%s-%s.%s.rpm' % (name, version, release, arch)
TypeError: %b requires a bytes-like object, or an object that implements 
__bytes__, not 'NoneType'
```

We could change build.py:

`canonname = rpmquery.RpmQuery.filename(self.mp['name'].encode(), epoch,  self.mp['version'].encode(), release, self.mp['arch'].encode())`

to
`canonname = rpmquery.RpmQuery.filename(self.mp['name'].encode(), epoch, self.mp['version'].encode(), release or b'0', self.mp['arch'].encode())`

marcus-h commented:
```
----- Forwarded message from Marcus Hüwe <suse-tux@gmx.de> -----
  
Date: Thu, 6 Feb 2020 20:23:26 +0100
From: Marcus Hüwe <suse-tux@gmx.de>
To: reply+AAIEUTBKRORG5ZJANIYOIVN3XPXHREVBNHHB4Y3OLA@reply.github.com
Subject: Re: [openSUSE/osc] handle empty release in rpmquery.py (#654)

On 2019-10-22 00:21:12 -0700, Marco Strigl wrote:
> We could change build.py:
> 
> `canonname = rpmquery.RpmQuery.filename(self.mp['name'].encode(), epoch,  self.mp['version'].encode(), release, self.mp['arch'].encode())`
> 
> to
> `canonname = rpmquery.RpmQuery.filename(self.mp['name'].encode(), epoch, self.mp['version'].encode(), release or b'0', self.mp['arch'].encode())`
> 
Yes, this change looks good (if the canonname is not used in "release is None"
case).

----- End forwarded message -----
```

lethliel commented:
Switched to

canonname = rpmquery.RpmQuery.filename(self.mp['name'].encode(), epoch,
self.mp['version'].encode(), release or b'0', self.mp['arch'].encode())

and removed the release handling in rpmquery.py
